s3/src/main/scala/com/meltwater/fawn/s3/S3.scala

```diff
@@ -0,0 +1,474 @@
+package com.meltwater.fawn.s3
+
+import cats.effect.concurrent.Ref
+import cats.effect.{Clock, ExitCase, Resource, Sync}
+import cats.implicits._
+import com.lucidchart.open.xtract.{ParseFailure, ParseSuccess, PartialParseSuccess, XmlReader}
+import com.meltwater.fawn.auth.V4Middleware
+import com.meltwater.fawn.common.{AWSCredentials, AWSRegion, AWSService}
+import org.http4s._
+import org.http4s.client.Client
+import org.http4s.implicits._
+import org.http4s.scalaxml._
+import org.typelevel.log4cats.Logger
+
+import scala.util.control.NoStackTrace
+import scala.xml.Elem
+
+/** Provides methods that correspond to AWS endpoints allowing interaction with S3 on AWS
+  */
+trait S3[F[_]] {
+  //Bucket Interaction
+  def createBucket(bucket: String, optHeaders: Option[Headers] = None): F[CreateBucketResponse]
+  def deleteBucket(bucket: String, optHeaders: Option[Headers] = None): F[GenericResponse]
+  def listBuckets(): F[ListBucketsResponse]
+
+  //Object Interaction
+  def listObjectsV2(bucket: String, optHeaders: Option[Headers] = None): F[ListObjectsResponse]
+  def putObject[T](bucket: String, key: String, t: T, optHeaders: Option[Headers] = None)(implicit
+      enc: EntityEncoder[F, T]): F[UploadFileResponse]
+  def getObject[T](bucket: String, key: String, optHeaders: Option[Headers] = None)(implicit
+      dec: EntityDecoder[F, T]): F[DownloadFileResponse[T]]
+  def deleteObject(
+      bucket: String,
+      key: String,
+      optHeaders: Option[Headers] = None): F[GenericResponse]
```

Comment:
we may want to revisit these later and properly model all the headers, but this is fine for the first version

Reply:
for now, I changed `Option[Headers]` to `Headers` since `Headers.empty` makes using `Option` rather pointless.